### ISSUE TITLE
Incidents not resolving

### ISSUE THREAD

**Issue by heffergm** [sensu-plugins/sensu-plugins-pagerduty#20]
As yet I can't replicate this 100% of the time. But often, I find that when I trigger an incident, when the condition that triggered said incident is resolved and the handler-pagerduty.rb says it resolved the incident, that in fact doesn't happen. I'm running sensu-plugins-pagerduty as installed by gem, version 0.0.9.

An example resolution issued here:

```
{"timestamp":"2016-02-27T18:23:33.733153+0000","level":"info","message":"handler output","handler":{"type":"pipe","command":"handler-pagerduty.rb","mutator":"pagerduty","name":"pagerduty"},"output":["/var/lib/gems/2.3.0/gems/sensu-plugin-1.2.0/lib/sensu-handler.rb:134:in `block in filter_silenced': Object#timeout is deprecated, use Timeout.timeout instead.\n/var/lib/gems/2.3.0/gems/sensu-plugin-1.2.0/lib/sensu-handler.rb:134:in `block in filter_silenced': Object#timeout is deprecated, use Timeout.timeout instead.\n/var/lib/gems/2.3.0/gems/sensu-plugin-1.2.0/lib/sensu-handler.rb:134:in `block in filter_silenced': Object#timeout is deprecated, use Timeout.timeout instead.\n/var/lib/gems/2.3.0/gems/sensu-plugins-pagerduty-0.0.9/bin/handler-pagerduty.rb:57:in `handle': Object#timeout is deprecated, use Timeout.timeout instead.\npagerduty -- Resolved incident -- routing3-valhalla_dev_us-east/routing3_cpu\n"]}
```

Left me with this issue still in pagerduty:

```
Details
Description
routing3-valhalla_dev_us-east/routing3_cpu : CheckCPU TOTAL WARNING: total=79.71 user=24.64 nice=0.0 system=4.93 idle=20.29 iowait=48.12 irq=0.0 s...
Details
timestamp     1456597053
action    create
occurrences   10
check     
total_state_change    5
history   ["0","0","0","0","0","0","0","0","0","0","0","1","1","1","1","1","1","1","1","1","1"]
status    1
output    CheckCPU TOTAL WARNING: total=79.71 user=24.64 nice=0.0 system=4.93 idle=20.29 iowait=48.12 irq=0.0 softirq=1.74 steal=0.29 guest=0.0
duration      2.069
executed      1456597050
issued    1456597050
name      routing3_cpu
refresh   3600
occurrences   10
interval      60
standalone    true
handlers      ["default"]
command   check-cpu.rb --sleep 2 -w 70 -c 90

client    
timestamp     1456597052
version   0.22.0
keepalive     
handle    true
refresh   1800
handlers      ["default"]
thresholds    
critical      600
warning   300


pd_override   
baseline      
warning   routing_low


pager_team    routing_low
address   192.168.2.94
name      routing3-valhalla_dev_us-east

id    024d99b8-431f-4d68-a09d-fa3918b07180
```

Any thoughts before I start delving into this more?

**Comment by analytically** [comment]
Did you set the incident_key? That's a requirement for resolving issues.

**Comment by scottcunningham** [comment]
+1, I'm seeing this now too. I worked around it by using revision d994264a3cd90ccd53621883b6c4b48c6b5099fc. Not sure if that's the latest working version for me but that's what was previously installed on our sensu box before I changed to using the ruby gem. Perhaps it's because of moving to the PagerDuty gem instead of Redphone?

**Comment by heffergm** [comment]
So the check is simply defined as part of a chef recipe in my case. I'm not doing any explicit setting of incident_key. Do I need to pass something into the blob now in order for resolutions to work? I'm not entirely clear on why this would be the case, given that sometimes I do see incidents resolving, but I haven't been able to determine as yet what the difference is between ones that work and ones that don't.

```
sensu_check "#{node[:hostname]}_cpu" do
  action cpu_action
  command "/usr/local/bin/check-cpu.rb \
--sleep 2 \
-w #{node[:mapzen_sensu_clients][:cpu_warn]} \
-c #{node[:mapzen_sensu_clients][:cpu_crit]}"
  handlers node[:mapzen_sensu_clients][:handlers_array]
  standalone true
  interval interval
  additional(occurrences: occurrences, refresh: refresh)
end
```

**Comment by scottcunningham** [comment]
Yeah, mine is part of a Puppet profile in my case so similar here.

**Comment by heffergm** [comment]
I'm using a setup where we're setting a warning baseline in the config to alert on a low urgency path, i.e.

```
    "pager_team": "ops",
    "pd_override": {
      "baseline": {
        "warning": "ops_low"
      }
    }
```

If I cause a condition where a pagerduty gets created at the warning level for ops_low, and then change that condition such that it becomes critical (but doesn't create a pagerduty incident), when the condition that triggered the incident is resolved sensu makes no attempt to resolve the incident in pagerduty... presumably since none exists for that pager_team.

I've also discovered that just getting incidents to resolve that were only triggered through one pager_team is hit or miss. Sometimes it'll work, sometimes not.

**Comment by heffergm** [comment]
Still drawing a blank here... I added some debug statements to the pagerduty handler. With the following incident triggered:

`Incident Key   sensu_sensu-server3-sensu_prod_us-east/sensu-server3_disk`

When resolved:

```
\nRESOLVING INCIDENT WITH INCIDENT KEY: sensu_sensu-server3-sensu_prod_us-east/sensu-server3_disk\nRESOLVING INCIDENT WITH DESCRIPTION: Sensu:  sensu-server3-sensu_prod_us-east/sensu-server3_disk : CheckDisk OK: All disk usage under 80% and inode usage under 85%\nRESOLVING INCIDENT WITH DETAILS: {\"id\"=>\"edb858fb-b5ed-4d93-a3aa-937bd6645011\", \"client\"=>{\"address\"=>\"10.0.21.15\", \"subscriptions\"=>[], \"keepalive\"=>{\"thresholds\"=>{\"warning\"=>300, \"critical\"=>600}, \"handlers\"=>[\"default\"], \"refresh\"=>1800, \"handle\"=>true}, \"name\"=>\"sensu-server3-sensu_prod_us-east\", \"pager_team\"=>\"ops_low\", \"pd_override\"=>{\"baseline\"=>{\"warning\"=>\"ops_low\"}}, \"version\"=>\"0.22.1\", \"timestamp\"=>1457646164}, \"check\"=>{\"command\"=>\"/usr/local/bin/check-disk-usage.rb -w 80 -c 90 -m .9\", \"standalone\"=>true, \"handlers\"=>[\"default\"], \"interval\"=>60, \"occurrences\"=>10, \"refresh\"=>3600, \"name\"=>\"sensu-server3_disk\", \"issued\"=>1457646177, \"executed\"=>1457646177, \"duration\"=>0.184, \"output\"=>\"CheckDisk OK: All disk usage under 80% and inode usage under 85%\\n\", \"status\"=>0, \"history\"=>[\"0\", \"0\", \"0\", \"0\", \"0\", \"0\", \"0\", \"0\", \"0\", \"0\", \"1\", \"1\", \"1\", \"1\", \"1\", \"1\", \"1\", \"1\", \"1\", \"1\", \"0\"], \"total_state_change\"=>11}, \"occurrences\"=>10, \"action\"=>\"resolve\", \"timestamp\"=>1457646177}\npagerduty -- Resolved incident -- sensu-server3-sensu_prod_us-east/sensu-server3_disk\n"]}
```

The incident keys match, but the incident is not resolved in pagerduty (not that the last log message is inaccurate: it's using `incident_key` rather than `incident_key_prefix` + `incident_key`)

**Comment by heffergm** [comment]
Happy days. I have a working regression.

The issue is that if the payload passed in the description greater than 1024 chars, the incident isn't resolved. This would appear to be a bug in the pagerduty api, as I'd expect a failure response.

```
A short description of the problem that led to this trigger. ... The maximum length is 1024 characters.
```

``` ruby
#!/usr/bin/env ruby

require 'pagerduty'

pagerduty = Pagerduty.new('your_service_key')

incident = pagerduty.trigger(
  "THIS IS A TEST",
  incident_key: "this_is_a_test"
)

arr = []
1024.times do
  arr << 'a'
end
details = arr.join('')

puts "Length of details: #{details.length}"
sleep 5

begin
  pagerduty.get_incident('this_is_a_test').resolve(
    details: details.slice(0..1024)
  )
rescue Net::HTTPServerException => error
  puts error.response.body
end
```

The above will create an incident successfully, but fail to resolve it.

Change `details: details.slice(0..1024)` to `details: details.slice(0..512)` and everything works as expected: the incident is created and resolved.

Any thoughts on how you'd prefer to fix this?

**Comment by analytically** [comment]
Create a PR with the slice 512.

**Comment by etherge** [comment]
Please refer issue #21 and PR #22.

**Comment by analytically** [comment]
Does PR #22 fix this issue as well?

**Comment by heffergm** [comment]
Looks like that's the idea. Can we get a new gem released and I'll kill my PR.

**Comment by heffergm** [comment]
Resolved in 1.0.0 release
